src/libraries/System.Linq/src/System/Linq/AnyAll.cs

```diff
@@ -32,6 +32,11 @@ public static bool Any<TSource>(this IEnumerable<TSource> source, Func<TSource,
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.predicate);
             }
 
+            if (source is List<TSource> list)
+            {
+                return list.Exists(element => predicate(element));
+            }
```

Comment:
As written, this is problematic because it causes a closure to be allocated on every use of Any, even when the source isn't a `List<TSource>`.  And when it is a `List<TSource>`, it results in an extra delegate invocation on every check.

The right way to do this currently is to use source.TryGetSpan to try to get a backing span from the source when it's either a `TSource[]` or a `List<T>`, and then just iterate through the resulting span.  I coincidentally have a change that does that (here and elsewhere in System.Linq) that I've been playing with over the weekend.  On the positive side, it makes cases like this one significantly faster, including avoiding the enumerator allocation.  On the negative side, when the type tests don't succeed, it adds a bit of overhead.

I'm still debating whether to open a PR with my changes.

Reply:
Very interesting, you're right that makes a huge difference! I'm happy to close this PR and leave it up to your WIP but just for my own sake I went ahead and tried out your suggestion: https://github.com/dotnet/runtime/pull/94983/commits/a93db4120204fe651c2478c807f1814da2cf3663

I saw the warning for the generic constraint -- did you plan to remove that in your WIP or did you take another approach to satisfy it?

Out of curiosity, what is it that's causing the span to be so much faster as opposed to a regular loop? Is it somehow applying SIMD under the hood if you iterate a span?

## Before

| Method                    | NumberOfItems | Mean          | Error      | StdDev     | Median        | Min           | Max           | Gen0   | Allocated |
|-------------------------- |-------------- |--------------:|-----------:|-----------:|--------------:|--------------:|--------------:|-------:|----------:|
| EnumerableAsListAny       | 10            |      25.13 ns |   0.547 ns |   0.537 ns |      24.90 ns |      24.52 ns |      26.06 ns | 0.0124 |     104 B |
| EnumerableAsArrayAny      | 10            |      24.13 ns |   0.246 ns |   0.231 ns |      24.17 ns |      23.71 ns |      24.52 ns | 0.0114 |      96 B |
| EnumerableAsEnumerableAny | 10            |      20.77 ns |   0.225 ns |   0.210 ns |      20.69 ns |      20.48 ns |      21.14 ns | 0.0124 |     104 B |
| EnumerableAsListAny       | 100           |     132.41 ns |   0.846 ns |   0.791 ns |     132.11 ns |     131.29 ns |     133.80 ns | 0.0123 |     104 B |
| EnumerableAsArrayAny      | 100           |     129.59 ns |   1.264 ns |   1.056 ns |     129.25 ns |     128.53 ns |     132.27 ns | 0.0110 |      96 B |
| EnumerableAsEnumerableAny | 100           |     123.91 ns |   1.282 ns |   1.071 ns |     124.01 ns |     122.48 ns |     126.50 ns | 0.0120 |     104 B |
| EnumerableAsListAny       | 100000        | 112,708.35 ns | 618.664 ns | 548.429 ns | 112,726.08 ns | 111,797.71 ns | 113,617.63 ns |      - |     104 B |
| EnumerableAsArrayAny      | 100000        | 100,290.80 ns | 633.322 ns | 561.423 ns | 100,154.08 ns |  99,431.25 ns | 101,519.82 ns |      - |      96 B |
| EnumerableAsEnumerableAny | 100000        |  94,320.25 ns | 483.748 ns | 428.830 ns |  94,329.21 ns |  93,634.17 ns |  95,040.19 ns |      - |     104 B |

## After

| Method                    | NumberOfItems | Mean         | Error      | StdDev     | Median       | Min          | Max          | Gen0   | Allocated |
|-------------------------- |-------------- |-------------:|-----------:|-----------:|-------------:|-------------:|-------------:|-------:|----------:|
| EnumerableAsListAny       | 10            |     12.12 ns |   0.233 ns |   0.207 ns |     12.09 ns |     11.73 ns |     12.55 ns | 0.0076 |      64 B |
| EnumerableAsArrayAny      | 10            |     12.52 ns |   0.223 ns |   0.197 ns |     12.50 ns |     12.25 ns |     12.87 ns | 0.0076 |      64 B |
| EnumerableAsEnumerableAny | 10            |     19.93 ns |   0.304 ns |   0.254 ns |     19.82 ns |     19.70 ns |     20.51 ns | 0.0124 |     104 B |
| EnumerableAsListAny       | 100           |     57.44 ns |   0.957 ns |   0.896 ns |     57.59 ns |     55.36 ns |     58.94 ns | 0.0075 |      64 B |
| EnumerableAsArrayAny      | 100           |     63.13 ns |   1.091 ns |   0.967 ns |     63.00 ns |     61.96 ns |     65.30 ns | 0.0074 |      64 B |
| EnumerableAsEnumerableAny | 100           |    122.12 ns |   0.749 ns |   0.701 ns |    121.92 ns |    121.17 ns |    123.58 ns | 0.0123 |     104 B |
| EnumerableAsListAny       | 100000        | 47,306.00 ns | 229.324 ns | 203.290 ns | 47,308.98 ns | 46,926.14 ns | 47,674.98 ns |      - |      64 B |
| EnumerableAsArrayAny      | 100000        | 36,189.09 ns | 545.226 ns | 510.004 ns | 36,047.41 ns | 35,627.41 ns | 37,268.59 ns |      - |      64 B |
| EnumerableAsEnumerableAny | 100000        | 93,606.80 ns | 814.899 ns | 722.387 ns | 93,354.19 ns | 92,955.80 ns | 95,292.45 ns |      - |     104 B |

## Summary

summary:
better: 7, geomean: 2.003
total diff: 7

No Slower results for the provided threshold = 2% and noise filter = 0.3 ns.

| Faster                                                                  | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| ----------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Linq.Tests.Perf_Any.EnumerableAsArrayAny(NumberOfItems: 100000)  |      2.78 |        100154.08 |         36047.41 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsListAny(NumberOfItems: 100000)   |      2.38 |        112726.08 |         47308.98 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsListAny(NumberOfItems: 100)      |      2.29 |           132.11 |            57.59 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsListAny(NumberOfItems: 10)       |      2.06 |            24.90 |            12.09 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsArrayAny(NumberOfItems: 100)     |      2.05 |           129.25 |            63.00 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsArrayAny(NumberOfItems: 10)      |      1.93 |            24.17 |            12.50 |         |
| System.Linq.Tests.Perf_Any.EnumerableAsEnumerableAny(NumberOfItems: 10) |      1.04 |            20.69 |            19.82 |         |

Reply:
> Out of curiosity, what is it that's causing the span to be so much faster as opposed to a regular loop?

It's primarily what I mentioned here:
"And when it is a `List<TSource>`, it results in an extra delegate invocation on every check."
Every element of the list required invoking the delegate you supplied which in turn needed to invoke the predicate delegate.

Reply:
That certainly clarifies why the span-based approach is better than the `.Exists()` variation, but what about the difference between Span and baseline? Is it effectively https://stackoverflow.com/a/63777097/1864167?

> Span guarantees a ["contiguous regions of arbitrary memory"](https://learn.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay) which allows the compiler to make optimizations to the CLI instructions.

Reply:
"baseline" being what's in main? That's going through IEnumerable, so every access requires going through the list's enumerator's MoveNext and Current.

Reply:
Got it, makes sense. I'll close this PR then and leave the actual change up to you, unless you think there's a slice of value that I can bring separately from what you're working on?